File: examples/sequential-cpp/example4.cpp
/* This code adds two vectors */

#include <iostream>
#include "utils/cppheader.h"

using namespace std;

const int SIZE = 100000000;

class AddVectors {
private:
	int *a, *b, *c;
	int size;

public:
	AddVectors(int *arrayC, int *arrayA, int *arrayB, int s) 
		: c(arrayC), a(arrayA), b(arrayB), size(s) {}

	void calculate() {
		for (int i = 0; i < size; i++) {
			c[i] = a[i] + b[i];
		}
	}
};

int main(int argc, char* argv[]) {
	Timer t;
	double ms;
	
	int *a = new int[SIZE];
	fill_array(a, SIZE);
	display_array("a", a);
	
	int *b = new int[SIZE];
	fill_array(b, SIZE);
	display_array("b", b);
	
	int *c = new int[SIZE];
	fill_array(c, SIZE);

	AddVectors av(c, a, b, SIZE);
	cout << "Starting..." << endl;
	ms = 0;
	for (int i = 0; i < N; i++) {
		t.start();
		av.calculate();
		ms += t.stop();
	}
	display_array("c", c);
	cout << "avg time = " << (ms/N) << " ms" << endl;
	
	delete [] a;
	delete [] b;
	delete [] c;
	
	return 0;
}

File: examples/fork-join/images/Example4.java
/* This code adds all the values of an array */

public class Example4 {
	private int a[], b[], c[];
	
	public Example4(int c[], int a[], int b[]) {
		this.a = a;
		this.b = b;
		this.c = c;
	}
	
	public void calculate() {
		for (int i = 0; i < c.length; i++) {
			c[i] = a[i] + b[i];
		}
	}
	
	public static void main(String args[]) {
		final int SIZE = 100_000_000;
		long startTime, stopTime;
		double acum = 0;
		
		int a[] = new int[SIZE];
		Utils.fillArray(a);
		Utils.displayArray("a", a);
		
		int b[] = new int[SIZE];
		Utils.fillArray(b);
		Utils.displayArray("b", b);
		
		int c[] = new int[SIZE];
		
		Example4 e = new Example4(c, a, b);
		acum = 0;
		for (int i = 0; i < Utils.N; i++) {
			startTime = System.currentTimeMillis();
			e.calculate();
			stopTime = System.currentTimeMillis();
			acum += (stopTime - startTime);
		}
		Utils.displayArray("c", c);
		System.out.printf("avg time = %.5f\n", (acum / Utils.N));
	}
}
			
